docs/internals/stream-corking.md

```diff
@@ -0,0 +1,254 @@
+
+The main tradeoff of `_writev` is in code complexity.  A stream can implement `_write`, `_writev`, or both.  It's up to the client programmer to ensure each of these methods behaves is a similar/reasonable way.
+
+This article won't get too into _how_ the client object uses `_write` and `_writev` --- we're more inserted in how stream corking delays calls to these methods.
+
+## Stream Corking and Agent Initialization
+
+To understand to first use case of stream corking in the client we need to understand how the client is instantiated, as well as understand the metadata that the  agent/client needs to collect before it can send data to APM Server.
+
+The APM Server APIs requires that each request begins with a metadata object.  This metadata contains information about the running agent instance.  Collecting some of this data, like the service name, is as simple as inspecting the agent's configuration.  Collecting data other data, however, requires us to query the local server environment in an asynchronous fashion.
+
+This means the client can't send data to APM Server right away.  The client currently uses stream corking to buffer data in the client until these asynchronous network requests are finished.
+
+Specifically, in the client's constructor function, [we `cork` the stream](https://github.com/elastic/apm-nodejs-http-client/blob/b992ef5a0a1eba3a2690cd275acb88938174c8fa/index.js#L100), make our asynchronous network request [via the `_fetchAndEncodeMetadata` function](https://github.com/elastic/apm-nodejs-http-client/blob/b992ef5a0a1eba3a2690cd275acb88938174c8fa/index.js#L101), and then [uncork the stream once this request has finished](https://github.com/elastic/apm-nodejs-http-client/blob/b992ef5a0a1eba3a2690cd275acb88938174c8fa/index.js#L106).
+
+```javascript
+  this.cork()
+  this._fetchAndEncodeMetadata(() => {
+    // _fetchAndEncodeMetadata will have set/memoized the encoded
+    // metadata to the _encodedMetadata property.
+
+    // this uncork reverse the .cork call in the constructor (above)
+    this.uncork()
+
+    // the `cloud-metadata` event allows listeners to know when the
+    // agent has finished fetching and encoding its metadata for the
+    // first time
+    this.emit('cloud-metadata', this._encodedMetadata)
+  })
+```
+
+ We also [_refuse_ to uncork the stream](https://github.com/elastic/apm-nodejs-http-client/blob/b992ef5a0a1eba3a2690cd275acb88938174c8fa/index.js#L350) until this metadata is present in the `_encodedMetadata` property.
+
+The effectively blocks data from being written to the stream chopper until these network requests finish.  Once the `_encodedMetadata` property is set, this specific instance of stream corking will not be invoked again.
+
+## Data Buffering
+
+The second way we use corking is to buffer our stream writes. Data is never immediately written to the stream chopper object.  Instead, the client stream will repeatedly
```

Comment:
The PR (https://github.com/elastic/apm-nodejs-http-client/pull/13) that added this said it is to improve performance without more details. I gather it is to bias to using _writev (for which the writable stream mechanics inside node are somewhat better), rather than _write.

Perhaps worth clarifying this "to buffer our stream writes", is about buffering writes (a) on the Client stream, before (b) writing them in bulk to the StreamChopper.